extensions/iceberg/s3/src/main/java/io/deephaven/extensions/s3/DeephavenAwsClientFactory.java

```diff
@@ -28,77 +28,85 @@ public class DeephavenAwsClientFactory implements AwsClientFactory, S3FileIOAwsC
-                DeephavenAwsClientFactory.class.getName()) != null) {
-            throw new IllegalArgumentException(
-                    String.format("Trying to put '%s', but it already exists", AwsProperties.CLIENT_FACTORY));
-        }
-        if (propertiesOut.putIfAbsent(S3FileIOProperties.CLIENT_FACTORY,
-                DeephavenAwsClientFactory.class.getName()) != null) {
-            throw new IllegalArgumentException(
-                    String.format("Trying to put '%s', but it already exists", S3FileIOProperties.CLIENT_FACTORY));
-        }
+        put(propertiesOut, AwsProperties.CLIENT_FACTORY, DeephavenAwsClientFactory.class.getName());
+        put(propertiesOut, S3FileIOProperties.CLIENT_FACTORY, DeephavenAwsClientFactory.class.getName());
         final String uuid = UUID.randomUUID().toString();
-        if (propertiesOut.putIfAbsent(UUID_KEY, uuid) != null) {
-            throw new IllegalArgumentException(
-                    String.format("Trying to put '%s', but it already exists", UUID_KEY));
-        }
-        propertiesOut.put(S3FileIOProperties.PRELOAD_CLIENT_ENABLED, "true");
+        put(propertiesOut, UUID_KEY, uuid);
+        // We are enabling preload to ensure that #initialize gets called during the creation of the catalog while we
+        // know the instructions are in the map.
         // Note: glue client is already preloaded on init if needed
+        propertiesOut.put(S3FileIOProperties.PRELOAD_CLIENT_ENABLED, "true");
         S3_INSTRUCTIONS_MAP.put(uuid, instructions);
         return () -> S3_INSTRUCTIONS_MAP.remove(uuid);
     }
 
+    private static <K, V> void put(Map<K, V> map, K key, V value) {
+        if (map.putIfAbsent(key, value) != null) {
+            throw new IllegalArgumentException(String.format("Key '%s' already exist in map", key));
+        }
+    }
+
     private static final Map<String, S3Instructions> S3_INSTRUCTIONS_MAP = new ConcurrentHashMap<>();
 
     private S3Instructions instructions;
 
+    @Override
+    public void initialize(Map<String, String> properties) {
+        final String uuid = properties.get(UUID_KEY);
+        if (uuid == null) {
+            throw new IllegalArgumentException(
+                    "DeephavenAwsClientFactory was setup improperly; it must be configured with DeephavenAwsClientFactory.addToProperties");
+        }
+        final S3Instructions s3i = S3_INSTRUCTIONS_MAP.get(uuid);
+        if (s3i == null) {
+            throw new IllegalStateException("This DeephavenAwsClientFactory was already cleaned up");
+        }
+        this.instructions = s3i;
+    }
+
+    private void checkInit() {
+        if (instructions == null) {
+            throw new IllegalStateException("Must initialize before use");
+        }
+    }
+
     @Override
     public S3Client s3() {
+        checkInit();
         return S3ClientFactory.getSyncClient(instructions);
     }
 
     @Override
     public GlueClient glue() {
```

Comment:
Are the glue, kms and dynamodb clients creation covered under any tests?

Reply:
No; you could argue we should prefer to throw unsupported for DynamoDB / KMS, but it was easy enough construct those clients if some Iceberg catalog happens to use them (I'm not sure under what circumstances, but it _is_ part of that API...).

I don't know if we've done any Glue testing. https://docs.localstack.cloud/user-guide/aws/glue/ might be possible with the "pro" version, not sure what that entails. https://docs.localstack.cloud/user-guide/aws/glue/

Reply:
I've done some manual testing of the Glue Catalog with the DeephavenAwsClientFactory, and it works. It was a good exercise though, because I realized that there is a new S3Client built for *ever* Table load (at least for the Glue catalog), so we can't immediately invoke the cleanup Runnable. After a quick change, it worked.

The only caller I see of kms() is LakeFormationAwsClientFactory - it's possible that some private Catalogs use it, but I'm not sure. It seems like this factory has something do do with Role ARN, and it may not actually be in use anymore since I think Iceberg has expanded how configurable AWS is since the creation of it.

The only caller of dynamodb() is DynamoDbCatalog; it may be possible to test it out with Localstack since it looks to be freely available, but it's just a matter of how much time do we want to spend trying? https://docs.localstack.cloud/user-guide/aws/dynamodb/

Reply:
Thanks for checking into this.
At this point, maybe just add a comment for kms and dynamodb?

Reply:
I added some comments in these implementations. Not sure it's worth it to elevate to a TODO / issue until we know it is necessary.

Reply:
Cool.